### ISSUE TITLE
Option to disable trivially_copy_pass_by_ref in public functions

### ISSUE THREAD

**Issue by mati865** [rust-lang/rust-clippy#4504]
Running Clippy over Rust repo yields about 100 unique warnings from `trivially_copy_pass_by_ref`, almost all of them were tripped by public functions which are not going to change.

Having option to disable it for public functions would help if this and probably other cases.

EDIT: `ptr_arg` have similar issue (but there are only 13 occurrences). Maybe something more generic should be developed?

**Comment by flip1995** [comment]
Just realized, this is kind of complicated without mentoring to do, so here we go:

Case 1:
https://github.com/rust-lang/rust-clippy/blob/7810652310a1e8d6cea6098b1b92e793e577d587/clippy_lints/src/trivially_copy_pass_by_ref.rs#L139-L147

with `cx.tcx.hir().get(hir_id)` the [`Node`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/hir/enum.Node.html) of the function in the `hir` can be accessed. This should be a `Node::Item(item)`, where the visibility can be accessed by `item.vis`

Case 2:
https://github.com/rust-lang/rust-clippy/blob/7810652310a1e8d6cea6098b1b92e793e577d587/clippy_lints/src/trivially_copy_pass_by_ref.rs#L129-L137

Here the visibility is dependent on the trait visibility. A trait in the `hir` is again an `Item`. This should be accessible by `cx.tcx.hir().get_parent_item(item.hir_id)`, which results in a `HirId`, which then should be resolvable with `cx.tcx.hir().get(_)`. This should be a `Node::Item(trait)`. The visibility then again is accessible with `trait.vis`.

One may also add sanity checks for the `ItemKind`s of the `Item`s returned by `cx.tcx.hir().get(_)`.

---
After this is implemented, a configuration option needs to be added. There should be enough examples in the code on how to do this.

**Comment by flip1995** [comment]
If this doesn't exist already, a function `visibility(hir_id)` would be useful in utils.

**Comment by flip1995** [comment]
Just realized, that there is a function [`visibility`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyCtxt.html#method.visibility), which should be used here :smile:

**Comment by palash25** [comment]
Hi @flip1995 new to Rust and I wanted to take a crack at this. I saw your comment explaining what needs to be done but its marked outdated anad collapsed, any reason for that? Is that approach not to be followed anymore?

**Comment by flip1995** [comment]
Hey @palash25 See the comment (https://github.com/rust-lang/rust-clippy/issues/4504#issuecomment-528442724) above yours. This isn't necessary, since the rust compiler already provides a function for that.

**Comment by palash25** [comment]
Oh my bad :sweat_smile: got confused.

Just to be sure I understand your comments correctly, the desired approach is to use the `visibility` method to figure out whether the method is public or not and if it is based on the flag disable or enable the lint, right?

**Comment by flip1995** [comment]
Yes correct. `visibility` takes a `DefId`. In the current implementation you have a `HirId`, which you can convert with [`opt_local_def_id`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/hir/map/struct.Map.html#method.opt_local_def_id).

> based on the flag disable or enable the lint

Yes, I'd suggest to enable/disable the visibility check:

```rust
if allow_public_flag && is_public() {
    return;
}
```

**Comment by flip1995** [comment]
@palash25 I just realized, that this is even easier: 

You can just use the function [`cx.access_levels.is_exported()`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/middle/privacy/struct.AccessLevels.html#method.is_exported) for visibility.

**Comment by palash25** [comment]
thanks for your help @flip1995 will try to take a stab at this in a day or two

**Comment by palash25** [comment]
@flip1995 need a little help with this part `allow_public_flag` I saw clippy help and it mentions the following flags
```
    -W --warn OPT       Set lint warnings
    -A --allow OPT      Set lint allowed
    -D --deny OPT       Set lint denied
    -F --forbid OPT     Set lint forbidden
```

Do I have to somehow reuse allow flag (because it seems lint specific so doesn't make much sense to me) for public functions or implement a new flag for it?

**Comment by palash25** [comment]
Actually the clippy help text isn't very helpful I am not able to find out what could be the possible values for `OPT` I tried running something like `cargo clippy --allow mut_from_ref` but got an error. Is this documented in detail somewhere somewhere, any examples of flag usage?

**Comment by mati865** [comment]
@palash25 it's explained in the readme: https://github.com/rust-lang/rust-clippy#allowingdenying-lints
TL;DR the syntax is as follows: `cargo clippy -- -A clippy::lint_name`.

**Comment by flip1995** [comment]
Also this is not the place to look for this kind of configuration. 

See for example this lint, for how to use a configuration:
https://github.com/rust-lang/rust-clippy/blob/edd90473ec5ba29b9ae1ee706c982e7046a6e63e/clippy_lints/src/functions.rs#L88-L97

https://github.com/rust-lang/rust-clippy/blob/0583181d036834788f24a0e4060f09a6615b0868/clippy_lints/src/lib.rs#L538

And this for how to add a configuration option:
https://github.com/rust-lang/rust-clippy/blob/db233b0920bbb840e191623e5d388232530557fd/clippy_lints/src/utils/conf.rs#L133-L134

To test it, you have to create a `ui-toml` test, like this one:
https://github.com/rust-lang/rust-clippy/tree/master/tests/ui-toml/toml_trivially_copy
